orchagent/dash/dashpavalidationorch.cpp

```diff
@@ -0,0 +1,364 @@
+    auto vni = vni_map_.find(entry.vni);
+    if (vni == vni_map_.end())
+    {
+        new_entry = true;
+        vni_map_[entry.vni] = {{entry.address, 1}};
+    } else {
+        auto& vni_addresses = vni->second;
+        auto address = vni_addresses.find(entry.address);
+        if (address == vni_addresses.end())
+        {
+            new_entry = true;
+            vni_addresses[entry.address] = 1;
+        } else {
+            vni_addresses[entry.address]++;
+        }
+
+        SWSS_LOG_DEBUG("PA Validation entry (%u, %s) refcount is increased to %" PRIu64, entry.vni, entry.address.to_string().c_str(), vni_addresses[entry.address]);
+    }
+
+    return new_entry;
+}
+
+bool DashPaValidationOrch::entryRefCountDec(const PaValidationEntry& entry)
+{
+    SWSS_LOG_ENTER();
+
+    auto vni = vni_map_.find(entry.vni);
+    if (vni == vni_map_.end())
+    {
+        SWSS_LOG_WARN("PA Validaion entry VNI %u IP %s is removed or not created yet", entry.vni, entry.address.to_string().c_str());
+        return false;
+    }
+
+    auto& vni_addresses = vni->second;
+    auto address = vni_addresses.find(entry.address);
+    if (address == vni_addresses.end())
+    {
+        SWSS_LOG_WARN("PA Validaion entry VNI %u IP %s is removed or not created yet", entry.vni, entry.address.to_string().c_str());
+        return false;
+    }
+
+    SWSS_LOG_DEBUG("PA Validation entry (%u, %s) refcount is decreased to %" PRIu64, entry.vni, entry.address.to_string().c_str(), address->second - 1);
+
+    if (address->second == 1)
+    {
+        vni_addresses.erase(address);
+        if (vni_addresses.empty())
+        {
+            vni_map_.erase(entry.vni);
+        }
+
+        return true;
+    } else {
+        address->second--;
+        return false;
+    }
+}
+
+bool DashPaValidationOrch::fetchVniAddresses(uint32_t vni, PaValidationEntryAddresses& addresses) const
+{
+    SWSS_LOG_ENTER();
+
+    auto it = vni_map_.find(vni);
+    if (it == vni_map_.end())
+    {
+        return true;
+    }
+
+    const auto& vni_addresses = it->second;
+
+    addresses.reserve(vni_addresses.size());
+    transform(vni_addresses.begin(), vni_addresses.end(), std::back_inserter(addresses),
+               [](const auto& kv) { return kv.first; });
+
+    return false;
+}
+
+void DashPaValidationOrch::paValidationConfigProcess(const PaValidationEntryList& toCreate, const PaValidationEntryList& toRemove)
+{
+    SWSS_LOG_ENTER();
+
+    auto task_status = paValidationConfigApply(toCreate, toRemove);
+    if (task_status != task_success)
+    {
+        parseHandleSaiStatusFailure(task_status);
+    }
+}
+
+task_process_status DashPaValidationOrch::paValidationConfigApply(const PaValidationEntryList& toCreate, const PaValidationEntryList& toRemove)
+{
+    EntityBulker<sai_dash_pa_validation_api_t> bulker(sai_dash_pa_validation_api, gMaxBulkSize);
+    std::vector<sai_status_t> statuses;
+
+    SWSS_LOG_ENTER();
+
+    statuses.reserve(toCreate.size() + toRemove.size());
```

Comment:
should we check PA validation CRM before calling bulk addition?

Reply:
The CRM checks are done by the CRM Orch, while the Orchs that are adding/removing the resources just update the usage. The API of CRM Orch:

https://github.com/sonic-net/sonic-swss/blob/1847195cf02a2baa478a664a671adc95c704a810/orchagent/crmorch.h#L73-L90